File: 0090-subsets-ii/0090-subsets-ii.cpp
class Solution {
private:
    void backtrack(int start, vector<int> &nums, vector<int> &current, vector<vector<int>> &result) {
        result.push_back(current);

        for (int i = start; i < nums.size(); ++i) {
            if (i > start && nums[i] == nums[i - 1]) {
                continue;
            }

            current.push_back(nums[i]);
            backtrack(i + 1, nums, current, result);
            current.pop_back();
        }
    }

public:
    vector<vector<int>> subsetsWithDup(vector<int>& nums) {
        vector<vector<int>> result;
        vector<int> current;

        sort(nums.begin(), nums.end());

        backtrack(0, nums, current, result);

        return result;
    }
};

File: 0090-subsets-ii/0090-subsets-ii.py
class Solution:
    def subsetsWithDup(self, nums: List[int]) -> List[List[int]]:
         # search in the tree
        # prevent the same number in the same level on the tree
        nums.sort()

        def backtrack(cur, start, result):
            result.append(cur[:])

            for i in range(start, len(nums)):
                if i > start and nums[i] == nums[i - 1]:
                    continue

                cur.append(nums[i])
                
                backtrack(cur, i + 1, result)

                cur.pop()

        result = []
        backtrack([], 0, result)

        return result


        
        